### PR TITLE
DAOS-9822 object: maintain RPC protocol bwd compatibility for object module

### PR DESCRIPTION
**Pull request by mchaarawi**
Signed-off-by: Mohamad Chaarawi <mohamad.chaarawi@intel.com>

### CODE REVIEW

File: src/object/obj_rpc.h [hunk 1]
```diff
@@ -63,9 +63,9 @@
 	X(DAOS_OBJ_RPC_PUNCH_AKEYS,					\
 		0, &CQF_obj_punch,					\
 		ds_obj_punch_handler, NULL, "akey_punch")		\
-	X(DAOS_OBJ_RPC_QUERY_KEY,					\
-		0, &CQF_obj_query_key,					\
-		ds_obj_query_key_handler, NULL, "key_query")		\
+	X(DAOS_OBJ_RPC_QUERY_KEY_0,					\
```
Comment (jolivier23) [inline src/object/obj_rpc.h:66]: I thought the whole point of the protocols was to actually have multiple versions so something like 

#define QUERY_KEY1 \
     X(....)
     
 #define QUERY_KEY2
     X(...)
     
 #define OBJ_PROTO_RPC_LIST(query_rpc)
   // replaces the query one with query_rpc so it's in same place in new protocol
  
  Anyway, I thought the intent of this stuff was to define one or the other and this should allow you to do that.  Not sure how difficult that is.

Comment (jolivier23) [reply src/object/obj_rpc.h:66]: You would need to be mindful of when you use the full macro and when you just use the query_key2 macro when defining stuff.

File: src/object/cli_mod.c [hunk 2]
```diff
@@ -43,23 +66,78 @@ dc_obj_init(void)
...
+	ep.ep_rank = 0;
+
+	rc = crt_proto_query_with_ctx(&ep, obj_proto_fmt_0.cpf_base, ver_array, 2, query_cb,
+				      &oproto, ctx);
+	if (rc) {
+		D_ERROR("crt_proto_query() failed: "DF_RC"\n", DP_RC(rc));
+		D_GOTO(out_grp, rc);
+	}
+
+	while (!oproto.completed) {
```
[the remark this answers is not in the document]
Comment (mchaarawi) [reply src/object/cli_mod.c:88]: i've dome something similar what we are doing on all daos RPCs:
https://github.com/daos-stack/daos/blob/master/src/client/api/rpc.c#L56

so if that's a problem we need to get it fixed there too..

File: src/object/cli_mod.c [hunk 3]
```diff
@@ -43,23 +66,78 @@ dc_obj_init(void)
...
+
+	rc = dc_mgmt_sys_attach(NULL, &sys);
+	if (rc != 0) {
+		D_ERROR("failed to attach to grp rc "DF_RC"\n", DP_RC(rc));
 		D_GOTO(out_class, rc);
 	}
 
+        ep.ep_grp = sys->sy_group;
+	ep.ep_tag = daos_rpc_tag(DAOS_REQ_MGMT, 0);
+	ep.ep_rank = 0;
```
[the remark this answers is not in the document]
Comment (mchaarawi) [reply src/object/cli_mod.c:79]: yes will add retry logic next

File: src/object/obj_rpc.h (hunk 1 above)
Comment (mchaarawi) [reply src/object/obj_rpc.h:66]: yes i was experimenting with this, but your idea should work. i'll push the change shortly

**Review by johannlombardi — APPROVED** [review]

**Review by jolivier23 — APPROVED** [review]

**Pull request merged**